models/src/main/protobuf/CasperMessage.proto

```diff
@@ -73,45 +25,126 @@ message DeployString {
   string term  = 3; //rholang source code to deploy (will be parsed into `Par`)
   bytes  sig   = 4; //signature of (hash(resource) + nonce) using private key
 }
-service DeployService {
-  rpc DoDeploy(DeployString) returns (DeployServiceResponse) {}
-  rpc createBlock(google.protobuf.Empty) returns (MaybeBlockMessage) {}
-  rpc addBlock(BlockMessage) returns (google.protobuf.Empty) {}
-  rpc showBlock(BlockQuery) returns (BlockInfo) {}
+
+message BlockQuery {
+  string hash = 1;
 }
+
 message DeployServiceResponse {
   bool   success = 1;
   string message = 2;
 }
+
 message MaybeBlockMessage {
   BlockMessage block = 1;
 }
 
-message Justification {
-  bytes validator       = 1;
-  bytes latestBlockHash = 2;
+message BlockQueryResponse {
+  string status = 1;
+  BlockInfo blockInfo = 2;
+}
+
+message BlocksResponse {
+  string status = 1;
+  repeated BlockInfo blocks = 2;
+  int64 length = 3;
+}
+
+// For node clients, see BlockMessage for actual Casper protocol Block representation
+message BlockInfo {
+  string blockHash = 1;
+  string blockSize = 2;
+  int64 blockNumber = 3;
+  int64 version = 4;
+  int32 deployCount = 5;
+  string tupleSpaceHash = 6; // Same as postStateHash of BlockMessage
+  string tupleSpaceDump = 7;
+  int64 timestamp = 8;
+  float faultTolerance = 9;
+  string mainParentHash = 10;
+  repeated string parentsHashList = 11;
+}
+
+// --------- End DeployService  --------
+
+// --------- Core Protocol  --------
+message BlockMessage {
+  bytes                  blockHash      = 1; //obtained by hashing the information in the header
+  Header                 header         = 2;
+  Body                   body           = 3;
+  repeated Justification justifications = 4; // map of all validators to latest blocks based on current view
+  bytes                  sender         = 5; //public key of the validator that created the block
+  bytes                  sig            = 6; //signature generated by signing `hash(hash(justification) concat blockHash)`.
+  string                 sigAlgorithm   = 7; //name of the algorithm used to sign
+  bytes reservedForFuture = 8;
 }
 
 message Header {
   repeated bytes parentsHashList = 1; //list of parent block hashes
   bytes postStateHash      = 2;
   bytes newCodeHash        = 3;
   bytes commReductionsHash = 4;
+  int64 timestamp = 5;
+  int64 version = 6;
```

Comment:
No but I've been convinced that we want to start out with a lot of extra bits wherever it could even possibly make sense and then get pressured to shrink.

Reply:
I was more wondering about the choice of data type than having a version at all. I think having a version is important for sure, but `Long` seems like a strange choice for the data type. I would have expected `String` (for say "1.0" or "mercury") or `Int` if it was just going to be purely numeric. I don't think we will have more versions than can fit in a 32-bit integer lol

Reply:
Oh yeah I was under the assumption you were thinking of int32. I didn't even think of String. I just inferred from https://insight.bitpay.com/api/block/000000000000000000170a0551c8f69e58d71efe2d8f2743adfdec780fd67dfc . I'm not sure why their version field is so long but as I mentioned above, when in doubt just keep it long for now.

Reply:
Oh I almost forgot, Bitcoin does a lot with their version field such as the first lowest X bits means this and the next Y bits means this, etc. You can search up the specifics.

Reply:
Hmm I think it makes sense. The alternative to deploy parallel soft/hard forks would be like

softFork1Flag = True
softFork2Flag = False
softFork3Flag = True

and you would have to have created those flags beforehand (or yet another fork to create those protobuf fields).